flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java

```diff
@@ -0,0 +1,296 @@
+					// if the initial submission failed. Never in case of a recovery
+					submittedJobGraphStore.removeJobGraph(jobId);
+				} catch (Throwable t) {
+					log.warn("Cannot remove job graph from submitted job graph store.", t);
+					e.addSuppressed(t);
+				}
+
+				throw new JobSubmissionException(jobId, "Could not start JobManager.", e);
+			}
+
+			jobManagerRunners.put(jobId, jobManagerRunner);
+
+			return Acknowledge.get();
+		} else {
+			throw new JobSubmissionException(jobId, "Job has already been submitted and " +
+				"is currently in state " + jobSchedulingStatus + '.');
+		}
+	}
+
+	@RpcMethod
+	Collection<JobID> listJobs() {
+		// TODO: return proper list of running jobs
+		return Collections.emptyList();
+	}
+
+	/**
+	 * Cleans up the job related data from the dispatcher. If cleanupHA is true, then
+	 * the data will also be removed from HA.
+	 *
+	 * @param jobId JobID identifying the job to clean up
+	 * @param cleanupHA True iff HA data shall also be cleaned up
+	 */
+	private void removeJob(JobID jobId, boolean cleanupHA) throws Exception {
+		JobManagerRunner jobManagerRunner = jobManagerRunners.remove(jobId);
+
+		if (jobManagerRunner != null) {
+			jobManagerRunner.shutdown();
+		}
+
+		if (cleanupHA) {
+			submittedJobGraphStore.removeJobGraph(jobId);
+		}
+
+		// TODO: remove job related files from blob server
+	}
+
+	protected abstract JobManagerRunner createJobManagerRunner(
+		ResourceID resourceId,
+		JobGraph jobGraph,
+		Configuration configuration,
+		RpcService rpcService,
+		HighAvailabilityServices highAvailabilityServices,
+		BlobService blobService,
+		HeartbeatServices heartbeatServices,
+		MetricRegistry metricRegistry,
+		OnCompletionActions onCompleteActions,
+		FatalErrorHandler fatalErrorHandler) throws Exception;
+
+	//------------------------------------------------------
+	// Utility classes
+	//------------------------------------------------------
+
+	private class DispatcherOnCompleteActions implements OnCompletionActions {
+
+		private final JobID jobId;
+
+		private DispatcherOnCompleteActions(JobID jobId) {
+			this.jobId = Preconditions.checkNotNull(jobId);
+		}
+
+		@Override
+		public void jobFinished(JobExecutionResult result) {
+			LOG.info("Job {} finished.", jobId);
+
+			runAsync(new Runnable() {
+				@Override
+				public void run() {
+					try {
+						removeJob(jobId, true);
+					} catch (Exception e) {
+						LOG.warn("Could not properly remove job {} from the dispatcher.", jobId, e);
+					}
+				}
+			});
+		}
+
+		@Override
+		public void jobFailed(Throwable cause) {
+			LOG.info("Job {} failed.", jobId);
+
+			runAsync(new Runnable() {
+				@Override
+				public void run() {
+					try {
+						removeJob(jobId, true);
+					} catch (Exception e) {
+						LOG.warn("Could not properly remove job {} from the dispatcher.", jobId, e);
+					}
+				}
+			});
+		}
+
+		@Override
+		public void jobFinishedByOther() {
+			LOG.info("Job {} was finished by other.", jobId);
```

Comment:
If there are multiple JobMasters (e.g. when having standby masters) and another JobMaster finished the job. For this implementation of the Dispatcher this won't be required, since we only start one JobMaster per job.